caddyconfig/httpcaddyfile/serveroptions.go

```diff
@@ -0,0 +1,247 @@
+						if err != nil {
+							return nil, d.Errf("parsing read_header timeout duration: %v", err)
+						}
+						serverOpts.ReadHeaderTimeout = caddy.Duration(dur)
+
+					case "write":
+						if !d.NextArg() {
+							return nil, d.ArgErr()
+						}
+						dur, err := caddy.ParseDuration(d.Val())
+						if err != nil {
+							return nil, d.Errf("parsing write timeout duration: %v", err)
+						}
+						serverOpts.WriteTimeout = caddy.Duration(dur)
+
+					case "idle":
+						if !d.NextArg() {
+							return nil, d.ArgErr()
+						}
+						dur, err := caddy.ParseDuration(d.Val())
+						if err != nil {
+							return nil, d.Errf("parsing idle timeout duration: %v", err)
+						}
+						serverOpts.IdleTimeout = caddy.Duration(dur)
+
+					default:
+						return nil, d.Errf("unrecognized timeouts option '%s'", d.Val())
+					}
+				}
+
+			case "max_header_size":
+				var sizeStr string
+				if !d.AllArgs(&sizeStr) {
+					return nil, d.ArgErr()
+				}
+				size, err := humanize.ParseBytes(sizeStr)
+				if err != nil {
+					return nil, d.Errf("parsing max_header_size: %v", err)
+				}
+				serverOpts.MaxHeaderBytes = int(size)
+
+			case "protocol":
+				for nesting := d.Nesting(); d.NextBlock(nesting); {
+					switch d.Val() {
+					case "allow_h2c":
+						if d.NextArg() {
+							return nil, d.ArgErr()
+						}
+						serverOpts.AllowH2C = true
+
+					case "experimental_http3":
+						if d.NextArg() {
+							return nil, d.ArgErr()
+						}
+						serverOpts.ExperimentalHTTP3 = true
+
+					case "strict_sni_host":
+						if d.NextArg() {
+							return nil, d.ArgErr()
+						}
+						trueBool := true
+						serverOpts.StrictSNIHost = &trueBool
+
+					default:
+						return nil, d.Errf("unrecognized protocol option '%s'", d.Val())
+					}
+				}
+
+			default:
+				return nil, d.Errf("unrecognized servers option '%s'", d.Val())
+			}
+		}
+	}
+	return serverOpts, nil
+}
+
+// validateServerOptions checks whether there's a conflict between
+// the configured global options and serverOptions, for backwards-compatibility
+func validateServerOptions(options map[string]interface{}) error {
+	// if we don't have "servers", then no errors to report
+	_, ok := options["servers"].([]serverOptions)
+	if !ok {
+		return nil
+	}
+
+	if _, ok := options["experimental_http3"]; ok {
+		return errors.New("cannot set 'experimental_http3' while also using 'servers'; please use the 'servers > protocol > experimental_http3' option instead")
+	}
+
+	return nil
+}
+
+// applyServerOptions sets the server options on the appropriate servers
+func applyServerOptions(servers map[string]*caddyhttp.Server, options map[string]interface{}) error {
+	// If experimental HTTP/3 is enabled, enable it on each server.
+	// We already know there won't be a conflict with serverOptions because
+	// we validated earlier that "experimental_http3" cannot be set at the same
+	// time as "servers"
+	if enableH3, ok := options["experimental_http3"].(bool); ok && enableH3 {
```

Comment:
Should we just deprecate that global option? (It is experimental, after all.) (To clarify, I mean to keep the one in the servers block, and have just one way to enable it.)

Reply:
Yeah, I figured we could mark it deprecated later but if you're happy with doing it now that works for me

Reply:
I usually prefer sooner rather than later before more people start using it.